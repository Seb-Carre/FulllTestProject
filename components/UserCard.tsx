import {
  Dimensions,
  Image,
  Linking,
  Pressable,
  StyleSheet,
  View,
} from 'react-native';

import Button from './Button';
import Checkbox from './Checkbox';
import CustomText from './CustomText';
import React from 'react';
import {User} from '../types';

type UserCardProps = {
  user: User;
  setSelected: (user: User) => void;
  selected: boolean;
};

/**
 * Functional Component to show the user card
 * @param {User} user
 * @param {(user: User) void} setSelected
 * Put the user to the selected list from the parent component
 * @param {boolean} selected
 * Manage the status of the checkbox inside the card
 */
export default function UserCard({user, setSelected, selected}: UserCardProps) {
  const onPressViewProfile = () => {
    // if the user has a html url
    if (user?.html_url) {
      // we are going to open this url in a web browser
      Linking.openURL(user.html_url);
    }
  };
  const onPressSelectedUser = () => {
    // using the setSelected method from parent component to manage the state
    setSelected(user);
  };
  return (
    <View style={styles.container}>
      <View style={styles.infosUserContainer}>
        <Pressable
          onPress={onPressSelectedUser}
          style={styles.checkboxPosition}>
          <Checkbox mode={selected ? 'active' : 'none'} />
        </Pressable>
        <View style={styles.infoUser}>
          <Image style={styles.tinyLogo} source={{uri: user?.avatar_url}} />
          <CustomText textType="title">{user?.id}</CustomText>
          <CustomText textType="title">{user?.login}</CustomText>
        </View>
      </View>
      <View style={styles.buttonContainer}>
        <Button onPress={onPressViewProfile}>View profile</Button>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: Dimensions.get('screen').width,
    borderRadius: 20,
    marginBottom: 16,
    backgroundColor: 'lightgray',
  },
  infosUserContainer: {flex: 1, padding: 20, flexDirection: 'row'},
  buttonContainer: {margin: 15, alignItems: 'center'},
  checkboxPosition: {
    position: 'absolute',
    padding: 25,
    zIndex: 1,
  },
  infoUser: {flex: 1, alignItems: 'center'},
  tinyLogo: {
    width: 70,
    height: 70,
    borderRadius: 35,
  },
});
