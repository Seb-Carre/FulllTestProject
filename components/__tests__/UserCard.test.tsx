import {fireEvent, render} from '@testing-library/react-native';

import {Linking} from 'react-native';
import React from 'react';
import {User} from '../../types';
import UserCard from '../UserCard';

const user: User = {
  login: 'sebastien',
  id: 404,
  avatar_url: 'https://reactnative.dev/img/tiny_logo.png',
  html_url: 'https://github.com/Seb-Carre',
};
const setSelectedMock = jest.fn();
const selected = true;
it('render the user card as snapShot expected', () => {
  const {toJSON} = render(
    <UserCard user={user} setSelected={setSelectedMock} selected={selected} />,
  );

  expect(toJSON()).toMatchSnapshot();
});

it('render the user card, check the user name and id', () => {
  const {queryByText} = render(
    <UserCard user={user} setSelected={setSelectedMock} selected={selected} />,
  );

  // We check if the login is showed
  expect(queryByText(user.login)).not.toBeNull();
  // We check if the ID is showed
  expect(queryByText(user.id.toString())).not.toBeNull();
});

it('render the user card and click on the checkBox', () => {
  const {getByTestId} = render(
    <UserCard user={user} setSelected={setSelectedMock} selected={selected} />,
  );

  // Making a press event and our mock function to be called
  fireEvent.press(getByTestId('UserCard.CheckBox' + user.id));
  expect(setSelectedMock).toBeCalled();
});

it('render the user card and click on view profile button', () => {
  // We are going to spy on the openURL Method from RN Linking
  const spy = jest.spyOn(Linking, 'openURL');
  const {getByTestId} = render(
    <UserCard user={user} setSelected={setSelectedMock} selected={selected} />,
  );

  fireEvent.press(getByTestId('UserCard.ViewProfileButton' + user.id));
  // We expect that the url is called is the url html.url to see his profile
  expect(spy).toHaveBeenCalledWith(user.html_url);
});
