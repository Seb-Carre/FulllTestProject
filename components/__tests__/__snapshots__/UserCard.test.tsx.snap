// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the user card as snapShot expected 1`] = `
<View
  style={
    {
      "backgroundColor": "lightgray",
      "borderRadius": 20,
      "height": 750,
      "marginBottom": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "padding": 25,
          "position": "absolute",
          "zIndex": 1,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#44a1f8",
            "borderRadius": 4,
            "borderWidth": 4,
            "height": 20,
            "justifyContent": "center",
            "width": 20,
          }
        }
        testID="UserCard.CheckBox404"
      >
        <Image
          source={
            {
              "testUri": "../../../assets/images/check-mark.png",
            }
          }
          style={
            {
              "height": 13,
              "width": 13,
            }
          }
          testID="Checkbox.icon"
        />
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Image
        source={
          {
            "uri": "https://reactnative.dev/img/tiny_logo.png",
          }
        }
        style={
          {
            "borderRadius": 35,
            "height": 70,
            "width": 70,
          }
        }
      />
      <Text
        style={
          [
            {
              "fontSize": 20,
              "fontWeight": "600",
            },
            undefined,
          ]
        }
      >
        404
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 20,
              "fontWeight": "600",
            },
            undefined,
          ]
        }
      >
        sebastien
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "margin": 15,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#00a2ff",
          "borderRadius": 10,
          "paddingHorizontal": 50,
          "paddingVertical": 20,
        }
      }
      testID="UserCard.ViewProfileButton404"
    >
      <Text
        style={
          [
            {},
            {
              "fontSize": 20,
            },
          ]
        }
      >
        View profile
      </Text>
    </View>
  </View>
</View>
`;
